File: September-LeetCoding-Challenge/04-Partition-Labels/Partition-Labels.cs
public class Solution {
    public IList<int> PartitionLabels(string S) {
        if(string.IsNullOrEmpty(S)) return null;
        
        int[] largestPositionByValue = new int[26];
        for(int i=0; i<S.Length; i++){
            largestPositionByValue[S[i] - 'a'] = i;
        }
        
        var res = new List<int>();
        var start = 0;
        var end=0;
        for(int i=0; i<S.Length; i++){
            end = Math.Max(end, largestPositionByValue[S[i]-'a']);
            if(end == i){
                res.Add(end-start+1);
                start = end + 1;
            }
        }
        return res;        
    }
}

File: September-LeetCoding-Challenge/04-Partition-Labels/Partition-Labels.rb
def partition_labels(s)
    last_indices = Array.new(26)
    s.each_char.with_index do |val, index|
       last_indices[val.ord - "a".ord] = index 
    end
    last_indices
    start = 0
    finish = 0
    out = []

    p last_indices
    0.upto((s.length)-1) do |n|
        val = s[n]
        finish = [finish, last_indices[val.ord - "a".ord]].max 
        if finish == n

          out << finish-start + 1
          start = finish  + 1
        end
    end
    out
end